packages/react-material-ui/src/components/Table/TableRoot.tsx

```diff
@@ -0,0 +1,220 @@
+ * @returns A React element representing the table root container.
+ */
+export const TableRoot = ({
+  children,
+  rows,
+  headers,
+  total,
+  pageCount,
+  tableQueryState: controlledTableQueryState,
+  updateTableQueryState: controlledUpdateTableQueryState,
+  ...rest
+}: PropsWithChildren<TableRootProps & BoxProps>) => {
+  const searchParams = useSearchParams();
+  const pathname = usePathname();
+  const router = useRouter();
+  const { tableQueryState, setTableQueryState } = useTableQueryState();
+
+  const [selected, setSelected] = useState<RowProps[]>([]);
+
+  const isControlled = !!controlledTableQueryState;
+  const handleUpdateTableQuery = isControlled
+    ? controlledUpdateTableQueryState
+    : setTableQueryState;
+
+  const { order, orderBy } = isControlled
+    ? controlledTableQueryState
+    : tableQueryState;
+
+  /**
+   * Handles the change of the number of rows displayed per page in the table.
+   *
+   * @param event - The event representing the row per page change.
+   */
+  const handleChangeRowsPerPage = (
+    event: React.ChangeEvent<HTMLInputElement>,
+  ) => {
+    const newRowsPerPageProperties = {
+      rowsPerPage: parseInt(event.target.value, 10),
+      page: 1,
+    };
+
+    handleUpdateTableQuery((prevState) => ({
+      ...prevState,
+      ...newRowsPerPageProperties,
+    }));
+
+    const newSearchParam = getSearchParams(
+      searchParams,
+      newRowsPerPageProperties,
+    );
+
+    if (newSearchParam) {
+      router.replace(`${pathname}?${newSearchParam}`);
+    }
+  };
+
+  /**
+   * Handles the selection of all checkboxes for rows in the table.
+   *
+   * @param event - The event representing the checkbox selection change.
+   */
+  const handleSelectAllCheckboxes = (
+    event: React.ChangeEvent<HTMLInputElement>,
+  ) => {
+    if (event.target.checked) {
+      setSelected(rows);
+      return;
+    }
+    setSelected([]);
+  };
+
+  /**
+   * Checks whether a row with the given ID is selected.
+   *
+   * @param id - The ID of the row to check.
+   * @returns `true` if the row is selected, `false` otherwise.
+   */
+  const isSelected = (id: string) =>
+    selected.findIndex((_row) => _row.id === id) !== -1;
+
+  /**
+   * Handles the selection of a checkbox for a specific row in the table.
+   *
+   * @param event - The mouse event triggering the checkbox selection.
+   * @param row - The row for which the checkbox is being selected.
+   */
+  const handleSelectCheckboxItem = (
+    event: React.MouseEvent<unknown>,
+    row: RowProps,
+  ) => {
+    const selectedIndex = selected.findIndex((_row) => _row.id === row.id);
+    let newSelected: RowProps[] = [];
+
+    if (selectedIndex === -1) {
+      newSelected = newSelected.concat(selected, row);
+    } else if (selectedIndex === 0) {
+      newSelected = newSelected.concat(selected.slice(1));
+    } else if (selectedIndex === selected.length - 1) {
+      newSelected = newSelected.concat(selected.slice(0, -1));
```

Comment:
This two is for performance?

Reply:
> Our goal is not to refactor every single line of code. Instead, we aim to simplify much of the existing code. You may still encounter old code that could be refactored in the future.


This code already exists in the codebase today and this also comes from the official docs, please check: https://mui.com/material-ui/react-table/#sorting-amp-selecting